pkg/injection/namespace/mapper/namespaces.go

```diff
@@ -4,38 +4,46 @@ import (
 	"context"
 
 	dynatracev1beta1 "github.com/Dynatrace/dynatrace-operator/pkg/api/v1beta1/dynakube"
+	injectionotel "github.com/Dynatrace/dynatrace-operator/pkg/injection/otel"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/otel/controller_runtime"
 	"github.com/pkg/errors"
 	corev1 "k8s.io/api/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 // NamespaceMapper manages the mapping creation from the namespace's side
 type NamespaceMapper struct {
-	ctx        context.Context
 	client     client.Client
 	apiReader  client.Reader
 	operatorNs string
 	targetNs   *corev1.Namespace
 }
 
-func NewNamespaceMapper(ctx context.Context, clt client.Client, apiReader client.Reader, operatorNs string, targetNs *corev1.Namespace) NamespaceMapper { //nolint:revive // argument-limit doesn't apply to constructors
-	return NamespaceMapper{ctx, clt, apiReader, operatorNs, targetNs}
+func NewNamespaceMapper(clt client.Client, apiReader client.Reader, operatorNs string, targetNs *corev1.Namespace) NamespaceMapper { //nolint:revive // argument-limit doesn't apply to constructors
+	return NamespaceMapper{controller_runtime.NewClient(clt), apiReader, operatorNs, targetNs}
 }
 
 // MapFromNamespace adds the labels to the targetNs if there is a matching Dynakube
-func (nm NamespaceMapper) MapFromNamespace() (bool, error) {
-	updatedNamespace, err := nm.updateNamespace()
+func (nm NamespaceMapper) MapFromNamespace(ctx context.Context) (bool, error) {
+	ctx, span := injectionotel.StartSpan(ctx, "NamespaceMapper.MapFromNamespace")
+	defer span.End()
+
+	updatedNamespace, err := nm.updateNamespace(ctx)
 	if err != nil {
 		return false, err
 	}
 	return updatedNamespace, nil
 }
 
-func (nm NamespaceMapper) updateNamespace() (bool, error) {
+func (nm NamespaceMapper) updateNamespace(ctx context.Context) (bool, error) {
+	ctx, span := injectionotel.StartSpan(ctx, "NamespaceMapper.updateNamespace")
```

Comment:
I am not an otel span expert, but I don't want to believe that we have to do it like this (to start a span for each function and close it), this feels unmaintainable and messy.

I would expect to start a span in the entry point, and end it there. (in the case of the webhook, when we get the request)
and maybe pass the span around aswell, to add something to it.

I looked at the keptn operator, their way makes more sense to me: https://github.com/keptn/lifecycle-toolkit/blob/main/lifecycle-operator/controllers/lifecycle/keptnworkloadversion/controller.go#L116

if you follow how they use that span, I hope it gets my point across.

Reply:
I'll check that, thx!

Reply:
It seems that they built a lot more framework/glue layer around OTel than I did in this PR. I tried to keep it as simple as possible on purpose to not completely overstep the focus of it. This PR has an demonstrator character, I think it's also good if the mechanics of OTel are still kept visible.

Passing around spans etc. additionally to a "tracer" context IMHO messes up the code and the function signatures much more than having a two liner to start a new span and attaching function relevant information there directly. In my mind, having one span per function also ties the resp. bits and pieces of information more tightly to the function that produces them. That being said, we still need to learn how we can actually use this information and if it really helps us.

Furthermore, they also start spans in lower parts of the code, just not on every level. It could be a feasible approach to only measure the "more interesting" or heavier/more expensive parts of the code and leave out smaller functions in between.

While I believe that there are good reasons for their way to do it I see a few drawbacks:
* to me it looks like they do not always end a span properly, which is a bit weird to me, not sure what negative side effects that could have on timings
* they do not get full call stack coverage which could reduce the insights, but that's something we need to consider
* they pass around additional contexts and objects, which really clutters the function signatures. From a quick glance I don't really understand, why they do it like this.